### PR TITLE
In listener detection, read listener_ip from self

### PR DESCRIPTION
**Pull request by williamjacksn**
`self.listener_ip` is read from the request argument and set on the object in `post()` before calls to `add_registered()` or `add_anonymous()`, so we don't need to call `get_argument()` over and over again.

I'm about to change the way `self.listener_ip` is detected so we can put Icecast behind a reverse proxy and still know the listener IP address, and I only want to change this in one place.

### CODE REVIEW

**Comment by williamjacksn** [comment]
I know you said you didn't like these changes, but I still think it's the best way to do it.

The changes in `fieldtypes.py` are for parsing the listener IP address out of the request that the Icecast server makes when authenticating a user. The `mount` argument used to be something like:

    /ocremix.mp3?3:1a2b3c4d5e

but with the reverse proxy I changed it to:

    /ocremix.mp3?3:1a2b3c4d5e&1.1.1.1

where `1.1.1.1` is the IP address of the listener.

The changes in `ldetect.py` save the listener IP address as an attribute of the Handler object and reuse it in all the places where it is needed.

**Comment by rmcauley** [comment]
Oh, yes, sorry for no reply - I just want to make sure we're not opening ourselves (or extending!) an SQL injection attack.  That field in the DB is just a string I think, and I forget what input checking I have around here.

**Comment by williamjacksn** [comment]
You're already using parameterized queries everywhere, so I don't think SQL injection is a possibility here.

**Comment by rmcauley** [comment]
Ah, yes, and we're already doing the validation in the fieldtypes.

Could we change to using e.g. `urlparse` in this fieldtype?  The regex you made is IPv4-only and now that I think about it, my original regex required the arguments to come in their specific order, which is a bit silly.

Would be good to parse the URL properly and get the parameters that way, and then use e.g. https://docs.python.org/3/library/ipaddress.html to see if it's a valid IP address instead of relying on a regex.  I can make this change if you don't want to! :)

**Comment by williamjacksn** [comment]
I extended the regex because that's what you already used. The current argument syntax (`?<user-id>:<key>`) doesn't follow the usual query parameter style.

Are you saying you want to start using something like `?user_id=3&key=abcde&ip=1.2.3.4` for listener detection?

**Comment by rmcauley** [comment]
I'm struggling here to be honest, with getting my head back into this code while maintaining perfect backwards and forwards compatibility, while not subjecting ourselves to garbage in the database, without typing or any automated testing...

Please bear with me.  Trying to wrap my head around this for the first time in a decade.  Let me write an essay on this, wooo!

### Weirdo URL Format

I completely forgot that I did not use real query strings and instead had this 1 string as the entire.  A voice in my brain is telling me not to use real query strings... but it's been at least a decade since I remember why.  It had something to do with Winamp 2 or IE7 not handling proper query URLs correctly?  Or maybe an old Icecast?  There was a reason for the weird schema in the first place.  Some popular client 15 years ago would pass along the first query parameter until an `&` was found.

### What Is That Format

Icecast sends us `listener_add/[station ID]?client=[icecast client ID]&ip=[end user IP]&mount=[mount]&agent=[end user HTTP agent]`

Without URL encoding, an example: `listener_add/1?client=124356&ip=2001:db8:3333:4444:5555:6666:7777:8888&mount=/game.ogg:1234:abcdefghijk&agent=Foobar 2.16`

### Database Implications

The `r4_listeners.ip` field which is used for anonymous voting has the type `TEXT`, not `INET`. Postgres is not saving my butt there if we don't do data validation before insertion.  Megabytes of text data can be slammed in.  On the RW API, the IP address that goes in here is provided by `X-Forwarded-For` through nginx and the Tornado Python lib I use in the API.

This is also the case for `client` (INT) and `agent` (TEXT).

I assume (😬) that's enough - the API whitelists these calls based on the `X-Forwarded-For` IP address, which unless our only singular relay is compromised is only said singular relay.

For the `ip` query parameter on the `listener_add` endpoint, because we whitelist the Icecast relays based on their origin IP, I'm trusting that the relay is an Icecast server sending us a sanitized IP address.

I don't remember any other ways an IP address can get in.

### Current Assumptions

I currently assume (but cannot prove!) that Rainwave as it exists does support IPv6, as all IP address related functions/validation are delegated to 3rd party programs/libs. (nginx, icecast, tornado, and TEXT fields in Pg)

I currently assume that the IP addresses provided are strings that have already been validated, and can only come from trusted 3rd parties, namely our API server's nginx and our whitelisted relay.

I currently assume that the worst that a forged IP address can do is enable someone to vote once per song per station per forged IP address.  This requires:
- Forging the IP for the relay-whitelisted `listener_add`
- Forging the IP through the RW API's nginx

### Change Goals

You know I mega-appreciate you and everything you've done for Rainwave.  I hope you know that too.  <3

I want to work together to make sure:

- Changes ensure IPv6 compatibility is retained
- Changes ensure the RW API server can trust that the IP address being provided to it are the end user's IP, is a valid IP address, and is the actual end user's IP address, without additional validation
- Changes ensure existing tune-in URLs must be continue working without modification, whether anonymous or tuned-in

### Oof

It's been so long since I've worked not only on this code base - in such a code base that exists without extensive unit and integration testing.  I've become used to professional, highly-automated-tested stacks, and this is very much "do it live.gif".. but then again, the stakes are low.

tl;dr: do it live.gif

**Comment by williamjacksn** [comment]
I wrote a massive comment describing the behavior and updated the code to use standard library parsing functions where possible. Please review and let me know what you think!
